core/drivers/rng_pta.c

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2021, Foundries Limited
+ */
+
+#include <crypto/crypto.h>
+#include <drivers/rng_pta_client.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <kernel/pseudo_ta.h>
+#include <kernel/spinlock.h>
+#include <kernel/timer.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+
+#define PTA_NAME "rng.pta"
+
+static TEE_Result rng_get_entropy(uint32_t types,
+				  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint8_t *e = NULL;
+	uint32_t rq_size = 0;
+	uint32_t exceptions;
+	TEE_Result res = TEE_SUCCESS;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	rq_size = params[0].memref.size;
+	e = (uint8_t *)params[0].memref.buffer;
+	if (!e)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	res = crypto_rng_read(e, rq_size);
+	thread_set_exceptions(exceptions);
+
+	return res;
+}
+
+static TEE_Result rng_get_info(uint32_t types,
+			       TEE_Param params[TEE_NUM_PARAMS])
+{
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/*
+	 * value.a is data rate: calls to crypto rng will always generate
+	 * the requested number of bytes from the RNG device therefore there
+	 * is no need to artificially wait for it to generate additional data.
+	 */
+	params[0].value.a = UINT32_MAX;
+
+	/*
+	 * value.b is quality: 0 (unknown), 1 (lowest) and 1024(highest)
+	 * HW dependent true entropy estimation set this to 1024.
+	 */
+	params[0].value.b = 1024;
```

Comment:
This should depend on a property on OP-TEE. Something like the Time protection level, hanlded in OP-TEE. See `tee_time_get_sys_time_protection_level()` for example.

Platforms with a full HW RNG, or a HW RNG seeding PRNG at each random genration request, could get level 1024, those seldom seeding PRNG with a HW RNG would get 128 and those without HW RNG would get 0, 1?

What do you think?

Reply:
I fully agree with needing to do something like that (not sure about the values themselves, will check the kernel driver).
Happy that you raised it! As a matter of fact I didnt post this patch earlier because I wasnt sure how to handle the different RNGs and platforms...do we need an additional interface in crypto_rng to get the quality?

Reply:
what about something along these lines?

```
diff --git a/core/include/crypto/crypto.h b/core/include/crypto/crypto.h
index 66ef2d204..999b279ac 100644
--- a/core/include/crypto/crypto.h
+++ b/core/include/crypto/crypto.h
@@ -314,6 +314,12 @@ enum crypto_rng_src {
        CRYPTO_RNG_SRC_NONSECURE        = (1 << 1 | 0),
 };

+enum crypto_rng_quality {
+ CRYPTO_RNG_IS_HW = 1 << 10,
+ CRYPTO_RNG_IS_HW_SEEDED = 1 << 7,
+ CRYPTO_RNG_IS_SW = 1 << 0,
+};
+
 /*
  * crypto_rng_init() - initialize the RNG
  * @data:      buffer with initial seed
@@ -347,6 +353,16 @@ void crypto_rng_add_event(enum crypto_rng_src sid, unsigned int *pnum,
  */
 TEE_Result crypto_rng_read(void *buf, size_t len);

+/*
+ * crypto_rng_info() - get the RNG quality information
+ *
+ * @quality:
+ *   CRYPTO_RNG_IS_HW        : HWRNG or PRNG with high frequency HWRNG seeding,
+ *   CRYPTO_RNG_IS_HW_SEEDED : PRNG with low frequency HWRNG seeding,
+ *   CRYPTO_RNG_IS_SW        : PRNG
+ */
+void crypto_rng_info(enum crypto_rng_quality *quality);
+
 /*
  * crypto_aes_expand_enc_key() - Expand an AES key
  * @key:       AES key buffer

```

Reply:
looks perfect to me